File: 11/nextpw.c
#include<stdio.h>
#include<string.h>

#define N 8
char pw [N];

int valid() {
  int i, n_pairs=0, increasing=0;
  char pair = '#';
  for (i=N-1; i>=0; i--) {
    if (pw[i] == 'i' || pw[i] == 'o' || pw[i] == 'l') {
      return 0;
    }
    if (n_pairs < 2 && i > 0) {
      if (pw[i] != pair && pw[i] == pw[i-1]) {
	pair = pw[i];
	n_pairs++;
      }
    }
    if (!increasing && i > 1) {
      if (pw[i] == pw[i-1]+1 && pw[i-1] == pw[i-2]+ 1) {
	increasing = 1;
      }
    }
  }
  if (n_pairs >= 2) {
    return increasing;
  } else {
    return 0;
  }
}

void incpw() {
  int i;
  for (i=N-1; i>=0; i--) {
    if (pw[i] >= 'z') {
      pw[i] = 'a';
    } else {
      pw[i] += 1;
      return;
    }
  }
}

int main(int argc, char** argv) {
  strcpy(pw, argv[1]);

  do {
    incpw();
  } while(!valid());
  printf("%s\n", pw);
}
  
  

File: 11/nextpw.py
import sys

forbidden = {'i', 'l', 'o'}

def valid(s):
    increasing = False
    pairs = set()
    for i in range(len(s)):
        if s[i] in forbidden:
            return False
        if i < len(s) - 1 and s[i] == s[i+1]:
            pairs.add(s[i])
        if not increasing and i < len(s) - 3:
            if ord(s[i]) + 1 == ord(s[i+1]) and ord(s[i+1]) + 1 == ord(s[i+2]):
                increasing = True
    return increasing and len(pairs) >= 2

def incpw(pw):
    newpw = pw
    for i in reversed(range(len(pw))):
        x = ord(pw[i])
        if x >= ord('z'):
            newpw = newpw[:i] + 'a' + newpw[i+1:]
        else:
            return newpw[:i] + chr(x+1) + newpw[i+1:]


def find_nextpw(pw):
    pw = sys.argv[1]

    nextpw = incpw(pw)
    while not valid(nextpw):
        nextpw = incpw(nextpw)
    print nextpw


if __name__ == "__main__":
    find_nextpw(sys.argv[1])
